include/alpaka/acc/Tag.hpp

```diff
@@ -0,0 +1,66 @@
+/* Copyright 2022 Simeon Ehrig
+ *
+ * This file is part of alpaka.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+#pragma once
+
+#include <alpaka/core/BoostPredef.hpp>
+
+#include <iostream>
+#include <type_traits>
+
+#define CREATE_ACC_TAG(tag_name)                                                                                      \
+    struct tag_name                                                                                                   \
+    {                                                                                                                 \
+        static std::string get_name()                                                                                 \
+        {                                                                                                             \
+            return #tag_name;                                                                                         \
+        }                                                                                                             \
+    }
+
+namespace alpaka
+{
+    CREATE_ACC_TAG(AccCpuFibersTag);
+    CREATE_ACC_TAG(AccCpuOmp2BlocksTag);
+    CREATE_ACC_TAG(AccCpuOmp2ThreadsTag);
+    CREATE_ACC_TAG(AccCpuSerialTag);
+    CREATE_ACC_TAG(AccCpuSyclIntelTag);
+    CREATE_ACC_TAG(AccCpuTbbBlocksTag);
+    CREATE_ACC_TAG(AccCpuThreadsTag);
+    CREATE_ACC_TAG(AccFpgaSyclIntelTag);
+    CREATE_ACC_TAG(AccFpgaSyclXilinxTag);
+    CREATE_ACC_TAG(AccGenericSyclTag);
+    CREATE_ACC_TAG(AccGpuCudaRtTag);
+    CREATE_ACC_TAG(AccGpuHipRtTag);
+    CREATE_ACC_TAG(AccGpuSyclIntelTag);
+    CREATE_ACC_TAG(AccOaccTag);
+    CREATE_ACC_TAG(AccOmp5Tag);
+
+    template<typename TAcc>
+    struct AccToTagTrait;
+
+    template<typename TAcc>
+    using AccToTag = typename AccToTagTrait<TAcc>::type;
+
+    template<typename TTag, typename TDim, typename TIdx>
+    struct TagToAccTrait;
+
+    template<typename TTag, typename TDim, typename TIdx>
+    using TagToAcc = typename TagToAccTrait<TTag, TDim, TIdx>::type;
+
+
+/* TODO: Remove the following pragmas once support for clang 6 is removed. They are necessary because
+        these /  clang versions incorrectly warn about a missing 'extern'. */
+#if BOOST_COMP_CLANG
+#    pragma clang diagnostic push
+#    pragma clang diagnostic ignored "-Wmissing-variable-declarations"
+#endif
+    template<typename TAcc, typename... TTag>
+    inline constexpr bool accMatchesTags = (std::is_same_v<AccToTag<TAcc>, TTag> || ...);
```

Comment:
A doxygen comment would be nice.

e.g.
```
Checks if the given accelerator is equal to one of the given tags in the list.
```

There should be no need for `inline`!

Reply:
Note: In PIConGPU we write such constructs always as list operations t allow the user to create an intermediate list of e.g. tags instead of hard code these into the call.

```C++
// this list is somewhere defined in the code, not directly where the tag matcher is used
using myPrefairedAccs = std::tupel<AccCpuOmp2BlocksTag, AccCpuSerialTag>;

auto isPrefaired = accMatchesTags<CurrentAcc,myPrefairedAccs>;
// we always allow passing one element only without saying it is a list, 
// in this case, we generate a list with one element and pass it to the 
// implementation which accepts lists only
auto isHip = accMatchesTags<CurrentAcc,AccGpuHipRtTag>;
```

Reply:
@psychocoderHPC Can you please point me to an example in PIConGPU, which we disused offline. I found an example in alpaka, but the requires also to pass the list type (e.g. `std::tuple`): https://github.com/alpaka-group/alpaka/blob/d98d7707a68e0adade3937b39c966cac9aad5e5d/include/alpaka/meta/CartesianProduct.hpp#L80-L81

Reply:
> There should be no need for `inline`!

The behaviour of `inline` changed with C++17. If we leave it out the compiler will complain about `accMatchesTag` being defined multiple times.

Reply:
> Note: In PIConGPU we write such constructs always as list operations t allow the user to create an intermediate list of e.g. tags instead of hard code these into the call.
> 
> ```c++
> // this list is somewhere defined in the code, not directly where the tag matcher is used
> using myPrefairedAccs = std::tupel<AccCpuOmp2BlocksTag, AccCpuSerialTag>;
> 
> auto isPrefaired = accMatchesTags<CurrentAcc,myPrefairedAccs>;
> // we always allow passing one element only without saying it is a list, 
> // in this case, we generate a list with one element and pass it to the 
> // implementation which accepts lists only
> auto isHip = accMatchesTags<CurrentAcc,AccGpuHipRtTag>;
> ```

I like the idea, but unfortunately we have no `boost mp11` and I didn't found any implementation in alpaka, which helps me to implement the feature. I would move the feature to a new issue/PR to get this PR finish soon.

Reply:
I am fine with adding this feature to another PR and not this.